metrics-core/src/main/java/com/codahale/metrics/LockFreeExponentiallyDecayingReservoir.java

```diff
@@ -0,0 +1,251 @@
+ * sampling reservoir, exponentially biased towards newer entries.
+ *
+ * @see <a href="http://dimacs.rutgers.edu/~graham/pubs/papers/fwddecay.pdf">
+ * Cormode et al. Forward Decay: A Practical Time Decay Model for Streaming Systems. ICDE '09:
+ * Proceedings of the 2009 IEEE International Conference on Data Engineering (2009)</a>
+ *
+ * {@link LockFreeExponentiallyDecayingReservoir} is based closely on the {@link ExponentiallyDecayingReservoir},
+ * however it provides looser guarantees while completely avoiding locks.
+ *
+ * Looser guarantees:
+ * <ul>
+ *     <li> Updates which occur concurrently with rescaling may be discarded if the orphaned state node is updated after
+ *     rescale has replaced it. This condition has a greater probability as the rescale interval is reduced due to the
+ *     increased frequency of rescaling. {@link #rescaleThresholdNanos} values below 30 seconds are not recommended.
+ *     <li> Given a small rescale threshold, updates may attempt to rescale into a new bucket, but lose the CAS race
+ *     and update into a newer bucket than expected. In these cases the measurement weight is reduced accordingly.
+ *     <li>In the worst case, all concurrent threads updating the reservoir may attempt to rescale rather than
+ *     a single thread holding an exclusive write lock. It's expected that the configuration is set such that
+ *     rescaling is substantially less common than updating at peak load. Even so, when size is reasonably small
+ *     it can be more efficient to rescale than to park and context switch.
+ * </ul>
+ *
+ * @author <a href="mailto:ckozak@ckozak.net">Carter Kozak</a>
+ */
+public final class LockFreeExponentiallyDecayingReservoir implements Reservoir {
+
+    private static final double SECONDS_PER_NANO = .000000001D;
+    private static final AtomicReferenceFieldUpdater<LockFreeExponentiallyDecayingReservoir, State> stateUpdater =
+            AtomicReferenceFieldUpdater.newUpdater(LockFreeExponentiallyDecayingReservoir.class, State.class, "state");
+
+    private final double alphaNanos;
+    private final int size;
+    private final long rescaleThresholdNanos;
+    private final Clock clock;
+
+    private volatile State state;
+
+    private final class State {
+        private final long startTick;
+        private final AtomicLong count;
+        private final ConcurrentSkipListMap<Double, WeightedSample> values;
+
+        State(long startTick, AtomicLong count, ConcurrentSkipListMap<Double, WeightedSample> values) {
+            this.startTick = startTick;
+            this.values = values;
+            this.count = count;
+        }
+
+        private void update(long value, long timestampNanos) {
+            double itemWeight = weight(timestampNanos - startTick);
+            double priority = itemWeight / ThreadLocalRandom.current().nextDouble();
+
+            long newCount = count.incrementAndGet();
+            if (newCount <= size || values.isEmpty()) {
```

Comment:
I don't understand why the isEmpty check is needed

Reply:
`count.incrementAndGet()` is invoked prior to adding elements to the map. Let's assume a small maximum reservoir size (though the same thing can happen with high concurrency):

Max is one
Two threads call histogram.update(value)
Thread 1: incrementAndGet returns 1 so `newCount <= size` is true, thread prepares to call values.put, but is descheduled (or thread 2 runs faster).

Thread 2: incrementAndGet returns 2 so `newCount <= size` is false. `values.firstKey()` returns `null` and  the comparison throws a NPE

Reply:
I suppose this could be restructured to avoid the `isEmpty` check and use a null check on `first` instead. What do you think?

Reply:
I don't think firstKey can return null, it would throw exception if the map is empty

Reply:
Maybe something like this should be used instead?
https://docs.oracle.com/javase/8/docs/api/java/util/NavigableMap.html#pollFirstEntry--

Reply:
You're right. In this case I think it must remain as written.

Reply:
> Maybe something like this should be used instead?
https://docs.oracle.com/javase/8/docs/api/java/util/NavigableMap.html#pollFirstEntry--

I don't think so, poll removes the entry which we only do if the priority is lower than our new element.

Reply:
Sorry, I don't mean to derail this PR too much, but I have thought a bit about this (since I also made that other similar PR).
Perhaps the code could be made both easier to reason about in terms of thread safety and more compact and readable if we first initialize the map to a map of exactly size dummy elements and then the update code is essentially just:

```java
if (values.putIfAbsent(priority, new WeightedSample(value, itemWeight)) == null) {
  values.pollFirstEntry();
}
```

That would maintain the map size invariant - when grabbing snapshot it would just filter out the dummy elements.

The dummy elements could have negative priority to ensure they are always evicted first. Real elements must have positive priority. The snapshot operation could filter out entries with negative weights.

Reply:
> Sorry, I don't mean to derail this PR too much, but I have thought a bit about this

I appreciate the feedback and ideas! This is the beauty of open source :-)

I like this your idea, I'll try to test it out!

Reply:
I tested using a slightly simplified implementation using the count, updated after a successful putIfAbsent:

```
private void update(long value, long timestampNanos) {
    double itemWeight = weight(timestampNanos - startTick);
    double priority = itemWeight / ThreadLocalRandom.current().nextDouble();

    if (values.putIfAbsent(priority, new WeightedSample(value, itemWeight)) == null
            && count.incrementAndGet() > size) {
        values.pollFirstEntry();
    }
}
```

Performance suffers compared to the original implementation:

```
Benchmark                                 (reservoirType)  Mode  Cnt      Score     Error  Units
ReservoirBenchmarks.updateReservoir            EXPO_DECAY  avgt    5   8620.807 ± 543.252  ns/op
ReservoirBenchmarks.updateReservoir  LOCK_FREE_EXPO_DECAY  avgt    5  20105.586 ± 982.248  ns/op
```

I think this is due to removal of the fast path in which no write operation is attempted on the map when the new priority does not exceed the minimum priority in the map.

Reply:
Right, maybe it would be possible to keep the fast path but still keep the size invariant

Reply:
I think you're right that we can optimize this using pollFirstEntry. Might be helpful to move `pollFirstEntry` into a small method to help escape analysis avoid an `AbstractMap.SimpleImmutableEntry` allocation.

```diff
// Always remove an item
- while (values.remove(first) == null) {
-     first = values.firstKey();
- }
+ values.pollFirstEntry();
```

Reply:
I made this branch: https://github.com/spkrka/metrics/tree/krka/poll-first-entry

Reply:
Figured out how to make a PR... https://github.com/carterkozak/metrics/pull/1/files